### PR TITLE
Baklava migration from VMs to Kubernetes

### PR DESCRIPTION
**Pull request by jcortejoso**
### Description

This PR contains the updates, fixes and changes needed for deploying the Baklava geth testnet (validators, proxies, fullnodes, private full-nodes) in Kubernetes.
PR forked from #6788.

### Other changes

- Removed old `.env` files.
- New env variable `PRIVATE_NODE_DISK_SIZE_GB` for specifying the disk size in the private (archive) txnodes (typically require more space compered to validators/proxies/fullnodes).
- allowVolumeExpansion to true in ssd storageClass #6513

### Tested

Used for the migration itself.

### Related issues

- Fixes celo-org/celo-labs#755 celo-org/celo-labs#766

### CODE REVIEW

File: .env.baklava [hunk 1]
```diff
@@ -30,11 +30,11 @@ CELOSTATS_TRUSTED_ADDRESSES=""
 CELOSTATS_BANNED_ADDRESSES=""
 CELOSTATS_RESERVED_ADDRESSES=""
 
-GETH_NODE_DOCKER_IMAGE_REPOSITORY="us.gcr.io/celo-org/geth"
-GETH_NODE_DOCKER_IMAGE_TAG="1.2.2-beta"
+GETH_NODE_DOCKER_IMAGE_REPOSITORY="us.gcr.io/celo-testnet/geth"
+GETH_NODE_DOCKER_IMAGE_TAG="f9c242ad250d018c805fb7ddebed13997c3d00a7"
```
Comment (oneeman) [inline .env.baklava]: I believe the commit to use is `dba095d2aea40e3ae5da65392de3b3b44105651b`.  `f9c242ad250d018c805fb7ddebed13997c3d00a7` is from the PR before it was merged.  But why do we need to use a commit hash at all?  Why not use `us.gcr.io/celo-org/geth` `1.2.2`, which seems to be the same as `dba095d2aea40e3ae5da65392de3b3b44105651b`

File: .env.baklava [hunk 2]
```diff
@@ -30,11 +30,11 @@ CELOSTATS_TRUSTED_ADDRESSES=""
...
 
 GETH_BOOTNODE_DOCKER_IMAGE_REPOSITORY="us.gcr.io/celo-testnet/geth-all"
-GETH_BOOTNODE_DOCKER_IMAGE_TAG="990a7a2b42ac8edbfb7389687c350baa657f583e"
+GETH_BOOTNODE_DOCKER_IMAGE_TAG="f9c242ad250d018c805fb7ddebed13997c3d00a7"
```
Comment (oneeman) [inline .env.baklava]: Same comment as above.

File: .env.baklava [hunk 3]
```diff
@@ -138,17 +135,20 @@ LOAD_TEST_TX_DELAY_MS=7200
 FAUCET_CUSD_WEI=60000000000000000000000
 
 # Baklava is run entirely by non-cLabs validators
-VALIDATORS=30
-VALIDATOR_PROXY_COUNTS=30:1
+VALIDATORS=2
+VALIDATOR_PROXY_COUNTS=2:1
```
Comment (oneeman) [inline .env.baklava]: ok for now, and once we're finishing the transition we'll want more varied proxy counts with at least some validators using multiple proxies

File: packages/helm-charts/testnet/templates/secondaries.statefulset.yaml [hunk 4]
```diff
@@ -182,7 +182,7 @@ spec:
           [[ "$IN_MEMORY_DISCOVERY_TABLE" == "true" ]] && ADDITIONAL_FLAGS="${ADDITIONAL_FLAGS} --use-in-memory-discovery-table"
 
           if [[ "{{ .Release.Name }}" == "alfajores" || "{{ .Release.Name }}" == "baklava" ]]; then
-            BOOTNODE_FLAG="--${network_name}"
+            BOOTNODE_FLAG="--{{ .Release.Name }}"
```
Comment (oneeman) [inline packages/helm-charts/testnet/templates/secondaries.statefulset.yaml]: Should the logic here match the new logic above? (which leaves out the bootnodes flag if this is mainnet)

File: .env [hunk 5]
```diff
@@ -102,7 +99,10 @@ VALIDATOR_PROXY_COUNTS=1:1,2:0
 STATIC_IPS_FOR_GETH_NODES=false
 # Whether tx_nodes/validators stateful set should use ssd persistent disks
 GETH_NODES_SSD_DISKS=true
+# Used for validators, proxies, and tx_nodes
 NODE_DISK_SIZE_GB=10
+# Used for private tx_nodes
```
Comment (oneeman) [inline .env]: Maybe clarify in the comment here and elsewhere that private tx nodes are archive nodes?

File: packages/celotool/src/lib/k8s-fullnode/base.ts [hunk 6]
```diff
@@ -101,7 +101,7 @@ export abstract class BaseFullNodeDeployer {
       `--set geth.metrics=${fetchEnvOrFallback(envVar.GETH_ENABLE_METRICS, 'false')}`,
       `--set genesis.networkId=${fetchEnv(envVar.NETWORK_ID)}`,
       `--set genesis.network=${this.celoEnv}`,
-      `--set geth.flags=${this.celoEnv === 'baklava' ? '--baklava' : ''}`,
+      // `--set geth.flags=${this.celoEnv === 'baklava' ? '--baklava' : ''}`,
```
Comment (oneeman) [inline packages/celotool/src/lib/k8s-fullnode/base.ts]: Remove rather than comment out

File: packages/helm-charts/common/templates/_helpers.tpl [hunk 7]
```diff
@@ -161,11 +168,17 @@ release: {{ .Release.Name }}
...
+      BOOTNODE_FLAG=""
+    else
+      BOOTNODE_FLAG="--bootnodes=$(cat /root/.celo/bootnodeEnode)"
+    fi
 {{ .extra_setup }}
-
     exec geth \
       --port $PORT  \
-      --bootnodes=$(cat /root/.celo/bootnodeEnode) \
+      "$BOOTNODE_FLAG" \
```
Comment (oneeman) [inline packages/helm-charts/common/templates/_helpers.tpl:187]: The `--alfajores` and `--baklava` flags (and the absence of one for mainnet) are not just for bootnodes.  They're for the bootnodes, network id, the genesis init step, hard fork activation blocks, and other configuration parameters which are hard-coded in geth.  I see you already conditioned the init step, so I think the thing that's left is moving the network id flag from always being given to also being part of this conditional.  I don't know whether that should be done using a separate variable that's set in the same if/elif/else, or a single variable that includes both the bootnode and the network id flags.

**Review by oneeman — COMMENTED** [review]
The Helm dry run parts should be reviewed by someone who understands that better.  But I reviewed the other changes and added some comments below

**Comment by oneeman** [comment]
For the final setup for Baklava, we should try to cover the following scenarios:

* Validator with no proxy
* Validator with 1 proxy
* Validator with 2 proxies
* Validator with >2 proxies
* Validator with 2 proxies and a secondary (should be doable using the work which was done in https://github.com/celo-org/celo-monorepo/pull/5029)

**Comment by nambrot** [comment]
Sorry for this drive-by comment, but very exciting!!!

File: .env.baklava (hunk 1 above)
Comment (jcortejoso) [reply .env.baklava]: Yes makes sense. I used the same docker tag that was being used in the VMs, but makes sense to use that one instead.

File: packages/helm-charts/common/templates/_helpers.tpl (hunk 7 above)
Comment (jcortejoso) [reply packages/helm-charts/common/templates/_helpers.tpl:187]: Added the networkId to the same code snippet.

**Comment by jcortejoso** [comment]
@oneeman Thanks for suggestions. Current validator-proxy configurations translates into:
`VALIDATOR_PROXY_COUNTS=2:0,2:1,1:2,1:3`.
I have to check how we can use the secondary nodes, and if we can select them arbitrary.

File: packages/helm-charts/testnet/templates/validators.statefulset.yaml [hunk 8]
```diff
@@ -90,7 +90,7 @@ spec:
           cat /root/.celo/bootnodeEnodeAddress
 
           [[ "$BOOTNODE_IP_ADDRESS" == 'none' ]] && BOOTNODE_IP_ADDRESS=${{ .Release.Namespace | upper }}_BOOTNODE_SERVICE_HOST
-          echo `cat /root/.celo/bootnodeEnodeAddress`@$BOOTNODE_IP_ADDRESS:30301 > /root/.celo/bootnodeEnode
+          echo "enode://$(cat /root/.celo/bootnodeEnodeAddress)@$BOOTNODE_IP_ADDRESS:30301" > /root/.celo/bootnodeEnode
```
Comment (trianglesphere) [inline packages/helm-charts/testnet/templates/validators.statefulset.yaml:93]: This change was not propogated on `secondaries.statefulset.yaml`

File: packages/helm-charts/testnet/templates/validators.statefulset.yaml [hunk 9]
```diff
@@ -196,6 +192,7 @@ spec:
           exec geth \
             $BOOTNODE_FLAG \
             --datadir /root/.celo \
+            --ipcpath=geth.ipc \
```
Comment (trianglesphere) [inline packages/helm-charts/testnet/templates/validators.statefulset.yaml:195]: ditto

**Review by trianglesphere — APPROVED** [review]
Helm and infra changes look good to me. One minor comment about how it's kinda hard to keep the charts for `validators` and `secondaries` in sync (they really only differ by a single command line option to the geth container).

**Comment by trianglesphere** [comment]
> I have to check how we can use the secondary nodes, and if we can select them arbitrary.

Re secondaries, they are configured with one secondary per validator with replicas for the first `m` validators where `m` is the number of replicas configured. Changing this could be done with some trickery about the replica ID how we assign proxies to validators.

File: packages/helm-charts/testnet/templates/validators.statefulset.yaml (hunk 8 above)
Comment (jcortejoso) [reply packages/helm-charts/testnet/templates/validators.statefulset.yaml:93]: updated

File: packages/helm-charts/testnet/templates/validators.statefulset.yaml (hunk 9 above)
Comment (jcortejoso) [reply packages/helm-charts/testnet/templates/validators.statefulset.yaml:195]: updated

**Comment by jcortejoso** [comment]
Everything is migrated now. Added 5 secondary nodes, bootnode ip address used in k8s is the same as it was using in the VM, and kubernetes deployment is in-sync with the code.
Not pending changes from my side here.

**Comment by oneeman** [comment]
Terrific, this will be a big relief for anyone working with Baklava!

**Pull request merged**